### ISSUE TITLE
Failed to infer the generic return type of class method.

### ISSUE THREAD

**Issue by huntzhan** [microsoft/pyright#3204]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**

The return type of class method is `Unknown` in the reported case.

**To Reproduce**
Steps to reproduce the behavior.

1. Copy the code snippet to VSCode.
2. Hover to show the type hint.

**Expected behavior**

`config_cls` should be inferred as `Type[T_DEMO_CONFIG]` instead of `Unknown`.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.

```python
from typing import get_args, Generic, Type, TypeVar


T_DEMO = TypeVar('T_DEMO', bound='Demo')
T_DEMO_CONFIG = TypeVar('T_DEMO_CONFIG')


class Demo(Generic[T_DEMO_CONFIG]):
    
    @classmethod
    def get_config_cls(cls: Type[T_DEMO]) -> Type[T_DEMO_CONFIG]:
        return get_args(cls.__orig_bases__[0])[0]  # type: ignore

    @classmethod
    def foo(cls: Type[T_DEMO]):
        config_cls = cls.get_config_cls()
```

Hovering on `get_config_cls` function definition looks good:

![image](https://user-images.githubusercontent.com/5213906/158548404-cc3e2951-3c21-4acf-8ed1-7f28fd24c1ff.png)

But hovering on the function invocation triggers the unexpected behavior:

![image](https://user-images.githubusercontent.com/5213906/158548435-3be15223-d265-4f4a-bf71-12d28b7cac12.png)

![image](https://user-images.githubusercontent.com/5213906/158548472-e3798c45-781f-4f13-9618-88436523a436.png)

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

I'm running Pyright as a VS Code extension, through Pylance version `v2022.3.1` (hence, Pyright version `1.1.228`)

**Additional context**

**Comment by erictraut** [comment]
I think the current behavior is correct. Mypy gives the same result.

You are using an incorrect combination of class-scoped and method-scoped type variables in your code. I'm struggling to understand the intent of your code, so I'm afraid I can't offer a suggested fix here.

**Comment by hmc-cs-mdrissi** [comment]
I'm guessing intent of code as I've written nearly same code before. My recommendation remove T_DEMO. T_DEMO_CONFIG is enough. This code hovers as expected,

T_DEMO_CONFIG = TypeVar("T_DEMO_CONFIG")


```
class Demo(Generic[T_DEMO_CONFIG]):
    @classmethod
    def get_config_cls(cls) -> Type[T_DEMO_CONFIG]:
        return get_args(cls.__orig_bases__[0])[0]  # type: ignore

    @classmethod
    def foo(cls):
        config_cls = cls.get_config_cls()

```

The issue with T_DEMO is T_DEMO is bound to a generic type and you are losing it's type argument. Mixing generic self typevars and also trying to use the inner type argument isn't doable. It's also not needed here.
